src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+          trace_msg(node, "Node has out-of-loop user found");
+          return false;
+        }
+      }
+    }
+  }
+  // 4) Bail out if the loop body has extra node
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* node = _body_nodes.at(idx);
+    if (!tracked.test(idx) && !in_core(node) && !is_memory_phi(node)) {
+      trace_msg(node, "Found extra loop node in loop body");
+      return false;
+    }
+  }
+  return true;
+}
+
+// Try creating a vector mask with the smallest vector element size
+const TypeVectMask* VectorMaskedLoop::create_vector_mask_type() {
+  BasicType vmask_bt = size_to_basic_type(_size_stats.smallest_size());
+  int vlen = Matcher::max_vector_size(vmask_bt);
+  if (!Matcher::match_rule_supported_vector(Op_LoopVectorMask, vlen, vmask_bt)) {
+    // Unable to create vector mask with the vlen & bt on this platform
+    return nullptr;
+  }
+  return (TypeVectMask*) TypeVect::makemask(vmask_bt, vlen);
+}
+
+// This checks if memory access node is our supported pattern
+bool VectorMaskedLoop::supported_mem_access(MemNode* mem) {
+  // First do a quick check by searching existing SWPointer(s)
+  for (int idx = 0; idx < _swptrs.length(); idx++) {
+    if (_swptrs.at(idx)->mem() == mem) {
+      return true;
+    }
+  }
+  // If not found, try creating a new SWPointer and insert it
+  SWPointer* ptr = mem_access_to_swpointer(mem);
+  if (ptr != nullptr) {
+    _swptrs.push(ptr);
+    return true;
+  }
+  return false;
+}
+
+// This tries creating an SWPointer object associated to the memory access.
+// Return nullptr if it fails or the SWPointer is not valid.
+SWPointer* VectorMaskedLoop::mem_access_to_swpointer(MemNode* mem) {
+  // Should access memory of a Java primitive value
+  BasicType mem_type = mem->memory_type();
+  if (!is_java_primitive(mem_type)) {
+    return nullptr;
+  }
+  // addp: memory address for loading/storing an array element. It should be an
+  // AddP node operating on an array of specific type
+  Node* addp = mem->in(MemNode::Address);
+  if (!addp->is_AddP() || !operates_on_array_of_type(addp, mem_type)) {
+    return nullptr;
+  }
+  // Create a Node_Stack for SWPointer's initial stack
+  Node_Stack* nstack = new Node_Stack(_arena, 5);
+  nstack->push(addp, 0);
+  // addp2: another possible AddP node for array element addressing. It should
+  // operate on the same memory type and have the same base with previous AddP.
+  Node* addp2 = addp->in(AddPNode::Address);
+  if (addp2->is_AddP()) {
+    if (!operates_on_array_of_type(addp2, mem_type) ||
+        addp->in(AddPNode::Base) != addp2->in(AddPNode::Base)) {
+      return nullptr;
+    }
+    nstack->push(addp2, 1);
+  }
+
+  // Check supported memory access via SWPointer. It's not supported if
+  //  1) The constructed SWPointer is invalid
+  //  2) Address is growing down (index scale * loop stride < 0)
+  //  3) Memory access scale is different from data size
+  //  4) The loop increment node is on the SWPointer's node stack
```

Comment:
Why should the `incr` not be on the node stack?

Reply:
That's a really corner case. In C2's ideal graph, most loop statements eventually uses the loop induction variable `phi` node as a input. That's good. But, there is one exception that a loop statement has a sub-expression of `iv + stride`. In this kind of cases, IGVN may do common sub-expression elimination and the inputs may come from the loop increment node thereafter. As the final step of vector masked transformation replaces the loop increment node, the calculation for `iv + stride` will also be replaced as well and it causes mis-compilation. In current patch, I duplicate the loop increment pattern for update (that's why we have `is_loop_incr_pattern()`, see commit 2) to avoid this issue, but currently it only applies to the expression not in array indices, such as `a[i] = i + 1`. For the patterns like `a[i+1] = i`, I'm still looking for a better approach to handle.

Reply:
Ok yes. You will have to find a way to separate out all these kinds of cases:
```
Using incr on backedge
Using incr as "i + stride" inside loop
Using incr outside/after loop -> final value + 1 -> count iterations
```

BTW: what happens if you use the `phi` directly after/outside the loop? Do you properly count the final value?